src/server/lock_manager/waiter_manager.rs

```diff
@@ -93,28 +83,32 @@ type Waiters = Vec<Waiter>;
 
 struct WaitTable {
     wait_table: HashMap<u64, Waiters>,
+    has_waiter: Arc<AtomicBool>,
 }
 
 impl WaitTable {
-    fn new() -> Self {
+    fn new(has_waiter: Arc<AtomicBool>) -> Self {
         Self {
             wait_table: HashMap::default(),
+            has_waiter,
         }
     }
 
-    #[allow(unused)]
+    #[cfg(test)]
     fn size(&self) -> usize {
         self.wait_table.iter().map(|(_, v)| v.len()).sum()
     }
 
-    fn set_wait_table_is_empty(&self) {
+    fn remove(&mut self, ts: u64) {
+        self.wait_table.remove(&ts);
         if self.wait_table.is_empty() {
-            store_wait_table_is_empty(true);
+            self.has_waiter.store(false, Ordering::Relaxed);
```

Comment:
Is it all right that `has_waiter` is not always accurate? Will it possibly cause a waiting transaction to be not waked up?

Reply:
I noted it at https://github.com/tikv/tikv/pull/5381/files/b5ea4d5068d3e8da21a32e22fdb145cfe9ebccca#diff-d76478cf71270ba36f4f48697f83ebdbR151.

I think it's acceptable because the waiting transaction will time out and the probability that the `has_watier` is not accurate is small.

Reply:
Got it. Thanks!